src/databricks/labs/ucx/assessment/dashboards.py

```diff
@@ -0,0 +1,370 @@
+from __future__ import annotations
+
+import itertools
+import json
+import logging
+from collections.abc import Iterable, Iterator
+from dataclasses import dataclass, field
+
+from databricks.labs.lsql.backends import SqlBackend
+from databricks.labs.lsql.lakeview import Dashboard as LsqlLakeviewDashboard, Dataset
+from databricks.sdk import WorkspaceClient
+from databricks.sdk.errors import DatabricksError
+from databricks.sdk.service.dashboards import Dashboard as SdkLakeviewDashboard
+from databricks.sdk.service.sql import Dashboard as SdkRedashDashboard, LegacyQuery
+
+from databricks.labs.ucx.framework.crawlers import CrawlerBase
+from databricks.labs.ucx.framework.utils import escape_sql_identifier
+
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class Query:
+    """UCX representation of a Query.
+
+    Note:
+        This class is not persisted into an inventory table. If you decide to persist this class, consider (future)
+        differences between Redash and Lakeview queries
+    """
+
+    id: str
+    """The ID for this query."""
+
+    name: str = "UNKNOWN"
+    """The title of this query that appears in list views, widget headings, and on the query page."""
+
+    parent: str = "ORPHAN"
+    """The identifier of the workspace folder containing the object."""
+
+    query: str = ""
+    """The text of the query to be run."""
+
+    catalog: str = ""
+    """The name of the catalog to execute this query in."""
+
+    schema: str = ""
+    """The name of the schema to execute this query in."""
+
+    tags: list[str] = field(default_factory=list)
+    """The tags set on this dashboard."""
+
+    @classmethod
+    def from_legacy_query(cls, query: LegacyQuery) -> Query:
+        """Create query from a :class:LegacyQuery"""
+        assert query.id
+        catalog = schema = None
+        if query.options:
+            catalog = query.options.catalog
+            schema = query.options.schema
+        return cls(
+            id=query.id,
+            name=query.name or cls.name,
+            parent=query.parent or cls.parent,
+            query=query.query or cls.query,
+            catalog=catalog or cls.catalog,
+            schema=schema or cls.schema,
+            tags=query.tags or [],
+        )
+
+    @classmethod
+    def from_lakeview_dataset(cls, dataset: Dataset, *, parent: str | None = None) -> Query:
+        """Create query from a :class:Dataset"""
+        return cls(
+            id=dataset.name,
+            name=dataset.display_name or cls.name,
+            parent=parent or cls.parent,
```

Comment:
The same comment applies here: if these values may not be available then we should make the fields optional rather than filling them with class-defined sentinels or some other default value.

Reply:
Sorry, missed those, thanks for catching!